### DISCUSSION TITLE
Why whitespace is been removed after wrap?

### DISCUSSION THREAD

**Discussion by linrongbin16** [mgeisler/textwrap#572] [Q&A] [answered]
Hi, thanks to your work, I have a simple question:

For example, we have the origin line: `"Hello World"` (length: 11), after using `textwrap::wrap(&line, 10)` method, the results are:

```rust
["Hello", "World"]
```

But why the whitespace `" "` (between `"Hello"` and `"World"`) is been removed? Is there any options that I could keep the origin text not been changed?

**Comment by mgeisler** [comment, marked as answer]
Good question! The whitespace at the ends of the lines is removed since it's normally unwanted.

The idea is that each line now takes up _at most_ 10 columns. If the trailing whitespace was left there, then this would not be true.

It goes further than this: the whitespace between the words is attached to the left word, and it is removed at the end of lines. So running

```rust
pub fn main() {
    dbg!(textwrap::wrap("This   is   a    demo   ", 10));
}
```

gives me

```
[/app/example.rs:2:5] textwrap::wrap("This   is   a    demo   ", 10) = [
    "This   is",
    "a    demo",
]
```

See https://godbolt.org/z/Kn9GTPvoM for an example of this (the normal Playground gives me an error right now).

What happens is that the text is split into _words_ using the [`WordSeparator`](https://docs.rs/textwrap/latest/textwrap/enum.WordSeparator.html). The default looks for Unicode break points, but you can also just split on ASCII space. The example shows you how the whitespace is attached to the left word:

```
let words = AsciiSpace.find_words("Hello World!").collect::<Vec<_>>();
assert_eq!(words, vec![Word::from("Hello "), Word::from("World!")]);
```

To keep the trailing whitespace, you would need to construct your own [`Word`s](https://docs.rs/textwrap/latest/textwrap/core/struct.Word.html). Unfortunately, I just now realize that it is not possible to easily do this because the `width` field is private...

You could implement [`Fragment`](https://docs.rs/textwrap/latest/textwrap/core/trait.Fragment.html) yourself, but then you need to call the low-level wrapping functions yourself.

Let me try and fix this for the next version!

**Reply by linrongbin16** [reply]
thanks for your detailed explain!
